flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/ChannelStateNoRescalingPartitionerTest.java

```diff
@@ -0,0 +1,129 @@
+import static org.apache.flink.runtime.checkpoint.StateObjectCollection.singleton;
+import static org.junit.Assert.fail;
+
+/**
+ * Simple test for
+ * {@link org.apache.flink.runtime.checkpoint.StateAssignmentOperation#channelStateNonRescalingRepartitioner channelStateNonRescalingRepartitioner}.
+ * It checks whether partitioner fails or not using property-based approach.
+ */
+@RunWith(Parameterized.class)
+public class ChannelStateNoRescalingPartitionerTest {
+
+	@Parameters
+	public static Collection<Object[]> parameters() {
+		List<Object[]> params = new ArrayList<>();
+		int[] parLevels = {1, 2};
+		int[] offsetSizes = {0, 1, 2};
+		final List<Function<OperatorSubtaskState, ?>> extractors = Arrays.asList(
+			OperatorSubtaskState::getInputChannelState,
+			OperatorSubtaskState::getResultSubpartitionState
+		);
+		for (int oldParallelism : parLevels) {
+			for (int newParallelism : parLevels) {
+				for (int offsetSize : offsetSizes) {
+					for (Function<OperatorSubtaskState, ?> stateExtractor : extractors) {
+						params.add(new Object[]{oldParallelism, newParallelism, offsetSize, stateExtractor});
+					}
+				}
+			}
+		}
+		return params;
+	}
+
+	private static final OperatorID OPERATOR_ID = new OperatorID();
+
+	private final int oldParallelism;
+	private final int newParallelism;
+	private final int offsetsSize;
+	private final Function<OperatorSubtaskState, ? extends StateObjectCollection<?>> extractState;
+
+	public ChannelStateNoRescalingPartitionerTest(int oldParallelism, int newParallelism, int offsetsSize, Function<OperatorSubtaskState, ? extends StateObjectCollection<?>> extractState) {
+		this.oldParallelism = oldParallelism;
+		this.newParallelism = newParallelism;
+		this.offsetsSize = offsetsSize;
+		this.extractState = extractState;
+	}
+
+	@Test
+	public <T extends AbstractChannelStateHandle<?>> void testNoRescaling() {
+		OperatorState state = new OperatorState(OPERATOR_ID, oldParallelism, oldParallelism);
+		state.putState(0, new OperatorSubtaskState(
+			empty(),
+			empty(),
+			empty(),
+			empty(),
+			singleton(new InputChannelStateHandle(new InputChannelInfo(0, 0), new EmptyStreamStateHandle(), getOffset())),
+			singleton(new ResultSubpartitionStateHandle(new ResultSubpartitionInfo(0, 0), new EmptyStreamStateHandle(), getOffset()))));
+		try {
+			// noinspection unchecked
+			StateAssignmentOperation.reDistributePartitionableStates(
+				singletonList(state),
+				newParallelism,
+				singletonList(OPERATOR_ID),
+				(Function<OperatorSubtaskState, StateObjectCollection<T>>) this.extractState,
+				channelStateNonRescalingRepartitioner("test"));
+		} catch (IllegalArgumentException e) {
+			if (!shouldFail()) {
+				throw e;
+			} else {
+				return;
+			}
+		}
+		if (shouldFail()) {
+			fail("expected to fail for: oldParallelism=" + oldParallelism + ", newParallelism=" + newParallelism + ", offsetsSize=" + offsetsSize + ", extractState=" + extractState);
+		}
+	}
```

Comment:
I think we could change the verify logic to 
~~~java
	@Test
	public <T extends AbstractChannelStateHandle<?>> void testNoRescaling() {
		OperatorState state = new OperatorState(OPERATOR_ID, oldParallelism, oldParallelism);
		state.putState(0, new OperatorSubtaskState(
			empty(),
			empty(),
			empty(),
			empty(),
			singleton(new InputChannelStateHandle(new InputChannelInfo(0, 0), new EmptyStreamStateHandle(), getOffset())),
			singleton(new ResultSubpartitionStateHandle(new ResultSubpartitionInfo(0, 0), new EmptyStreamStateHandle(), getOffset()))));
		verifyNoRescalingChannelState(state, OperatorSubtaskState::getInputChannelState);
		verifyNoRescalingChannelState(state, OperatorSubtaskState::getResultSubpartitionState);
	}

	private <T extends AbstractChannelStateHandle<?>> void verifyNoRescalingChannelState(OperatorState state, Function<OperatorSubtaskState, StateObjectCollection<T>> stateExtractor) {
		try {
			// noinspection unchecked
			StateAssignmentOperation.reDistributePartitionableStates(
				singletonList(state),
				newParallelism,
				singletonList(OPERATOR_ID),
				stateExtractor,
				channelStateNonRescalingRepartitioner("test"));
		} catch (IllegalArgumentException e) {
			if (!shouldFail()) {
				throw e;
			} else {
				return;
			}
		}
		if (shouldFail()) {
			fail("expected to fail for: oldParallelism=" + oldParallelism + ", newParallelism=" + newParallelism + ", offsetsSize=" + offsetsSize + ", stateExtractor=" + stateExtractor);
		}
	}
~~~

Reply:
All of the parameters are indeed parameters of the tested method.
I think it's better to keep their combinations in a single place so it's easy to see what combinations are tested.